Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -154,31 +161,108 @@ namespace ROS2
             pf.count = 1;
             message.fields.push_back(pf);
         }
 
-        size_t sizeInBytes = lastScanResults.m_points.size() * sizeof(AZ::Vector3);
+        if (lastScanResults.m_ids.has_value()) {
+            sensor_msgs::msg::PointField pfId;
+            pfId.name = "entity_id";
+            pfId.offset = message.point_step;
+            pfId.datatype = sensor_msgs::msg::PointField::INT32;
+            pfId.count = 1;
+            message.fields.push_back(pfId);
+            constexpr auto fieldLength = sizeof(int32_t);
+            message.point_step += fieldLength;
+        }
+        AZStd::array<AZ::Color, 256> colorLookupTable;
+        if (lastScanResults.m_classes.has_value()) {
+            colorLookupTable = m_lidarCore.m_lidarConfiguration.
+                    GenerateSegmentationColorsLookupTable();
+
+            AZStd::array<const char *, 3> colorFieldNames = {"r", "g", "b"};
+
+            for (int i = 0; i < colorFieldNames.size(); i++) {
+                sensor_msgs::msg::PointField pf;
+                pf.name = colorFieldNames[i];
+                pf.offset = message.point_step;
+                pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
+                pf.count = 1;
+                message.fields.push_back(pf);
+                message.point_step += sizeof(float);
+            }
+
+            sensor_msgs::msg::PointField pfClassId;
+            pfClassId.name = "class_id";
+            pfClassId.offset = message.point_step;
+            pfClassId.datatype = sensor_msgs::msg::PointField::UINT8;
+            pfClassId.count = 1;
+            message.fields.push_back(pfClassId);
+            message.point_step += sizeof(uint8_t);
+        }
+
+        const auto sizeInBytes = pointCount * message.point_step;
         message.data.resize(sizeInBytes);
+
+        PackedVector3 xyz = {0.0f, 0.0f, 0.0f};
+
+        // IMO this is the clearest way to handle the offset with future fields, compiler should optimize it
+        for (int i = 0; i < pointCount; ++i) {
+            // to avoid alignment issues, we copy the data field by field
+            xyz[0] = lastScanResults.m_points[i].GetX();
+            xyz[1] = lastScanResults.m_points[i].GetY();
+            xyz[2] = lastScanResults.m_points[i].GetZ();
+            memcpy(&message.data[i * message.point_step], &xyz, sizeof(PackedVector3));
+            uint32_t nextFieldOffset = sizeof(PackedVector3);
+
+            if (lastScanResults.m_ids.has_value()) {
+                memcpy(&message.data[i * message.point_step + nextFieldOffset], &lastScanResults.m_ids.value()[i],
+                       sizeof(int32_t));
+                nextFieldOffset += sizeof(int32_t);
+            }
+
+            if (lastScanResults.m_classes.has_value()) {
+                for (int j = 0; j < 3; j++) {
+                    float color = colorLookupTable[lastScanResults.m_classes.value()[i]].GetElement(j);
```

Comment:
Capture `colorLookupTable[lastScanResults.m_classes.value()[i]` before this for-loop?